ros/src/twist_controller/twist_controller.py

```diff
@@ -10,77 +12,112 @@
 
-        # if a < 0:
+        # keeping speed but minor error_v
         if 0.0 < error_v < 0.5:
-            # only correct when abs(error_v) greater than 0.5
             brake = 0.0
-            throttle = a * 0.75 if a > 0.0 else 0.0
+            throttle = min(self._max_throttle, a) * 0.75 if a > 0.0 else 0.0
         elif error_v <= 0 or twist_linear_x < 0:
-            brake = -a * self.total_mass * self.wheel_radius
+            brake = -a * self._total_mass * self._wheel_radius
 
-            # keep braking as long as twist velocity < 0
-            if brake > 0.0 and twist_linear_x < 0:
+            # keep braking while not moving, e.g. traffic lights
+            if brake > 0.0 and twist_linear_x < 0.0 and current_velocity_x < 5.0:
                 brake = BrakeCmd.TORQUE_MAX * 0.5
-            elif brake < self.deadband_torque:
+            elif brake < self._deadband_torque:
                 brake = 0.0
             throttle = 0.0
 
             rospy.logdebug("twist_linear_x={}\terror_v={}\tbrake={}".format(twist_linear_x, error_v, brake))
 
         else:
             brake = 0.0
-            throttle = min(1.0, a)
-
-        self.twist_values[0].append(twist_linear_x)
-        self.twist_values[1].append(current_velocity_x)
-        self.twist_values[2].append(brake)
-        self.twist_values[3].append(throttle)
-        self.twist_values[4].append(error_v)
-        self.twist_values[5].append(self.low_pass_filter.get())
-
-        steer = self.yaw_controller.get_steering(twist_linear_x, twist_angular_z, current_velocity_x)
+            throttle = min(self._max_throttle, a)
+
+        self._twist_values[0].append(twist_linear_x)
+        self._twist_values[1].append(current_velocity_x)
+        self._twist_values[2].append(brake)
+        self._twist_values[3].append(throttle)
+        self._twist_values[4].append(error_v)
+        self._twist_values[5].append(self._low_pass_filter.get())
+
+        steer = self._yaw_controller.get_steering(twist_linear_x, twist_angular_z, current_velocity_x)
+
+        self._log_data.append({
+            'twist_linear_x': twist_linear_x,
+            'current_velocity_x': current_velocity_x,
+            'error_v': error_v,
+            'throttle': throttle,
+            'brake': brake,
+            'steer': steer,
+            'twist_angular_z': twist_angular_z
+        })
 
         return throttle, brake, steer
 
     def reset(self):
-        self.pid.reset()
\ No newline at end of file
+        self._pid.reset()
+
+    def save(self):
+        basepath = os.path.dirname(os.path.abspath(__file__))
+        logfile = os.path.join(basepath, 'twist_controller.csv')
+        fieldnames = ['twist_linear_x', 'current_velocity_x', 'error_v', 'throttle', 'brake', 'steer', 'twist_angular_z']
+        with open(logfile, 'w') as csvfile:
+            writer = csv.DictWriter(csvfile, fieldnames=fieldnames)
+            writer.writeheader()
+            writer.writerows(self._log_data)
```

Comment:
You are writing a `twist_controller.csv` file here. I'm pretty sure this will also work on Carla, but we should be aware that we will never get access to that file from Carla. ;)